In [34]:
# import libraries
import pandas as pd

# Explore dataset

In [35]:
df = pd.read_csv("data\online_shoppers_intention.csv" )
df.head()

Administrative  Administrative_Duration  Informational  \
0             0.0                      0.0            0.0   
1             0.0                      0.0            0.0   
2             0.0                     -1.0            0.0   
3             0.0                      0.0            0.0   
4             0.0                      0.0            0.0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0             1.0                 0.000000   
1                     0.0             2.0                64.000000   
2                    -1.0             1.0                -1.000000   
3                     0.0             2.0                 2.666667   
4                     0.0            10.0               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0        1       1            1  Returning_Visitor    False    False  
1        2       1            2  Returning_Visitor    False    False  
2        1       9            3  Returning_Visitor    False    False  
3        2       2            4  Returning_Visitor    False    False  
4        3       1            4  Returning_Visitor     True    False

In [36]:
df.describe().transpose()

count         mean          std  min         25%  \
Administrative           12316.0     2.317798     3.322754  0.0    0.000000   
Administrative_Duration  12316.0    80.906176   176.860432 -1.0    0.000000   
Informational            12316.0     0.503979     1.270701  0.0    0.000000   
Informational_Duration   12316.0    34.506387   140.825479 -1.0    0.000000   
ProductRelated           12316.0    31.763884    44.490339  0.0    7.000000   
ProductRelated_Duration  12316.0  1196.037057  1914.372511 -1.0  185.000000   
BounceRates              12316.0     0.022152     0.048427  0.0    0.000000   
ExitRates                12316.0     0.043003     0.048527  0.0    0.014286   
PageValues               12330.0     5.889258    18.568437  0.0    0.000000   
SpecialDay               12330.0     0.061427     0.198917  0.0    0.000000   
OperatingSystems         12330.0     2.124006     0.911325  1.0    2.000000   
Browser                  12330.0     2.357097     1.717277  1.0    2.000000   
Region                   12330.0     3.147364     2.401591  1.0    1.000000   
TrafficType              12330.0     4.069586     4.025169  1.0    2.000000   

                                50%          75%           max  
Administrative             1.000000     4.000000     27.000000  
Administrative_Duration    8.000000    93.500000   3398.750000  
Informational              0.000000     0.000000     24.000000  
Informational_Duration     0.000000     0.000000   2549.375000  
ProductRelated            18.000000    38.000000    705.000000  
ProductRelated_Duration  599.766190  1466.479902  63973.522230  
BounceRates                0.003119     0.016684      0.200000  
ExitRates                  0.025124     0.050000      0.200000  
PageValues                 0.000000     0.000000    361.763742  
SpecialDay                 0.000000     0.000000      1.000000  
OperatingSystems           2.000000     3.000000      8.000000  
Browser                    2.000000     2.000000     13.000000  
Region                     3.000000     4.000000      9.000000  
TrafficType                2.000000     4.000000     20.000000

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12330 entries, 0 to 12329
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Administrative           12316 non-null  float64
 1   Administrative_Duration  12316 non-null  float64
 2   Informational            12316 non-null  float64
 3   Informational_Duration   12316 non-null  float64
 4   ProductRelated           12316 non-null  float64
 5   ProductRelated_Duration  12316 non-null  float64
 6   BounceRates              12316 non-null  float64
 7   ExitRates                12316 non-null  float64
 8   PageValues               12330 non-null  float64
 9   SpecialDay               12330 non-null  float64
 10  Month                    12330 non-null  object 
 11  OperatingSystems         12330 non-null  int64  
 12  Browser                  12330 non-null  int64  
 13  Region                   12330 non-null  int64  
 14  TrafficType           

1. Delete all null values [ONLY 14 rows out of 12,330 rows (0.11%), i.e., considered to be safe to delete all null values]
2. Transfer boolean values into numerical values
2. Need to transform "Month" and "VisitorType" to numeric

# Clean dataset
## 1. Delete all null values from the dateset

In [38]:
# Drop null values
df.dropna(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12316 entries, 0 to 12329
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Administrative           12316 non-null  float64
 1   Administrative_Duration  12316 non-null  float64
 2   Informational            12316 non-null  float64
 3   Informational_Duration   12316 non-null  float64
 4   ProductRelated           12316 non-null  float64
 5   ProductRelated_Duration  12316 non-null  float64
 6   BounceRates              12316 non-null  float64
 7   ExitRates                12316 non-null  float64
 8   PageValues               12316 non-null  float64
 9   SpecialDay               12316 non-null  float64
 10  Month                    12316 non-null  object 
 11  OperatingSystems         12316 non-null  int64  
 12  Browser                  12316 non-null  int64  
 13  Region                   12316 non-null  int64  
 14  TrafficType           

## 2. Transfer boolean in to numerical values

In [39]:
df.Weekend = df.Weekend.astype(int)
df.Revenue = df.Revenue.astype(int)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12316 entries, 0 to 12329
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Administrative           12316 non-null  float64
 1   Administrative_Duration  12316 non-null  float64
 2   Informational            12316 non-null  float64
 3   Informational_Duration   12316 non-null  float64
 4   ProductRelated           12316 non-null  float64
 5   ProductRelated_Duration  12316 non-null  float64
 6   BounceRates              12316 non-null  float64
 7   ExitRates                12316 non-null  float64
 8   PageValues               12316 non-null  float64
 9   SpecialDay               12316 non-null  float64
 10  Month                    12316 non-null  object 
 11  OperatingSystems         12316 non-null  int64  
 12  Browser                  12316 non-null  int64  
 13  Region                   12316 non-null  int64  
 14  TrafficType           

## 3. One hot Code "Month" and "VisitorType" columns

In [40]:
# check all the month values
print('Month: ', df['Month'].unique())

Month:  ['Feb' 'Mar' 'May' 'Oct' 'June' 'Jul' 'Aug' 'Nov' 'Sep' 'Dec']


In [41]:
# use pd.concat to join the new columns with the original dataframe
df = pd.concat([df,pd.get_dummies(df['Month'], prefix='Month')],axis=1)
df = pd.concat([df,pd.get_dummies(df['VisitorType'], prefix='VisitorType')],axis=1)

# drop the original 'Month' column (don't need it anymore)
df.drop(['Month', 'VisitorType'],axis=1, inplace=True)

Administrative  Administrative_Duration  Informational  \
0             0.0                      0.0            0.0   
1             0.0                      0.0            0.0   
2             0.0                     -1.0            0.0   
3             0.0                      0.0            0.0   
4             0.0                      0.0            0.0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0             1.0                 0.000000   
1                     0.0             2.0                64.000000   
2                    -1.0             1.0                -1.000000   
3                     0.0             2.0                 2.666667   
4                     0.0            10.0               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay  ...  Month_Jul  Month_June  \
0         0.20       0.20         0.0         0.0  ...          0           0   
1         0.00       0.10         0.0         0.0  ...          0           0   
2         0.20       0.20         0.0         0.0  ...          0           0   
3         0.05       0.14         0.0         0.0  ...          0           0   
4         0.02       0.05         0.0         0.0  ...          0           0   

   Month_Mar  Month_May  Month_Nov  Month_Oct  Month_Sep  \
0          0          0          0          0          0   
1          0          0          0          0          0   
2          0          0          0          0          0   
3          0          0          0          0          0   
4          0          0          0          0          0   

   VisitorType_New_Visitor  VisitorType_Other  VisitorType_Returning_Visitor  
0                        0                  0                              1  
1                        0                  0                              1  
2                        0                  0                              1  
3                        0                  0                              1  
4                        0                  0                              1  

[5 rows x 29 columns]

In [58]:
df.head()

Administrative  Administrative_Duration  Informational  \
0             0.0                      0.0            0.0   
1             0.0                      0.0            0.0   
2             0.0                     -1.0            0.0   
3             0.0                      0.0            0.0   
4             0.0                      0.0            0.0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0             1.0                 0.000000   
1                     0.0             2.0                64.000000   
2                    -1.0             1.0                -1.000000   
3                     0.0             2.0                 2.666667   
4                     0.0            10.0               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay  ...  Month_Jul  Month_June  \
0         0.20       0.20         0.0         0.0  ...          0           0   
1         0.00       0.10         0.0         0.0  ...          0           0   
2         0.20       0.20         0.0         0.0  ...          0           0   
3         0.05       0.14         0.0         0.0  ...          0           0   
4         0.02       0.05         0.0         0.0  ...          0           0   

   Month_Mar  Month_May  Month_Nov  Month_Oct  Month_Sep  \
0          0          0          0          0          0   
1          0          0          0          0          0   
2          0          0          0          0          0   
3          0          0          0          0          0   
4          0          0          0          0          0   

   VisitorType_New_Visitor  VisitorType_Other  VisitorType_Returning_Visitor  
0                        0                  0                              1  
1                        0                  0                              1  
2                        0                  0                              1  
3                        0                  0                              1  
4                        0                  0                              1  

[5 rows x 29 columns]

In [59]:
print('Administrative: ', df['Administrative'].unique())
print('Informational: ', df['Informational'].unique())
print('ProductRelated: ', df['ProductRelated'].unique())
print('PageValues: ', df['PageValues'].unique())
print('SpecialDay: ', df['SpecialDay'].unique())
print('OperatingSystems: ', df['OperatingSystems'].unique())
print('Browser: ', df['Browser'].unique())
print('Region: ', df['Region'].unique())
print('TrafficType: ', df['TrafficType'].unique())

Administrative:  [ 0.  1.  2.  4. 12.  3. 10.  6.  5.  9.  8. 16. 13. 11.  7. 18. 14. 17.
 19. 15. 24. 22. 21. 20. 23. 27. 26.]
Informational:  [ 0.  1.  2.  4. 16.  5.  3. 14.  6. 12.  7.  9. 10.  8. 11. 24. 13.]
ProductRelated:  [  1.   2.  10.  19.   3.  16.   7.   6.  23.  13.  20.   8.   5.  32.
   4.  45.  14.  52.   9.  46.  15.  22.  11.  12.  36.  42.  27.  90.
  18.  38.  17. 128.  25.  30.  21.  51.  26.  28.  31.  24.  50.  96.
  49.  68.  98.  67.  55.  35.  37.  29.  34.  71.  63.  87.  40.  33.
  54.  64.  75.  39. 111.  81.  61.  47.  44.  88. 149.  41.   0.  79.
  66.  43. 258.  80.  62.  83. 173.  48.  58.  57.  56.  69.  82.  59.
 109. 287.  53.  84.  78. 137. 113.  89.  65.  60. 104. 129.  77.  74.
  93.  76.  72. 194. 140. 110. 132. 115.  73. 328. 160.  86. 150.  95.
 130. 151. 117. 124. 127. 125. 116. 105.  92. 157. 154. 220. 187. 112.
 131. 159.  94. 204. 142. 206. 102. 313. 145.  85.  97. 198. 181. 126.
 106. 101. 108. 119.  70. 122.  91. 276. 100. 291. 114. 172